sdk-api-src/content/d3dcommon/nn-d3dcommon-id3ddestructionotifier.md

```diff
@@ -0,0 +1,108 @@
+old-location: direct3d11\id3dinclude.htm
+tech.root: direct3d11
+ms.assetid: 2020ce65-3a6e-4a9f-9e97-b94e3c75f4f5
+ms.date: 12/05/2018
+ms.keywords: ID3DDestructionNotifier, ID3DDestructionNotifier interface [Direct3D], ID3DDestructionNotifier interface [Direct3D],described, d3dcommon/ID3DDestructionNotifier, direct3d.id3ddestructionnotifier
+f1_keywords:
+- d3dcommon/ID3DDestructionNotifier
+dev_langs:
+- c++
+req.header: d3dcommon.h
+req.include-header: 
+req.target-type: Windows
+req.target-min-winverclnt: Windows 7 [desktop apps \| UWP apps]
+req.target-min-winversvr: Windows Server 2008 R2 [desktop apps \| UWP apps]
+req.kmdf-ver: 
+req.umdf-ver: 
+req.ddi-compliance: 
+req.unicode-ansi: 
+req.idl: 
+req.max-support: 
+req.namespace: 
+req.assembly: 
+req.type-library: 
+req.irql: 
+topic_type:
+- APIRef
+- kbSyntax
+api_type:
+- COM
+api_location:
+- D3DCommon.h
+api_name:
+- ID3DDestructionNotifier
+targetos: Windows
+req.typenames: 
+req.redist: 
+ms.custom: 19H1
+---
+
+# ID3DInclude interface
+
+
+## -description
+
+
+<b>ID3DDestructionNotifier</b> is an interface that can be used to register for callbacks when a D3D Nano-COM object is destroyed.
+To acquire an instance of this interface, call <a href="windows/win32/api/unknwn/nf-unknwn-iunknown-queryinterface(refiid_void)"></a> on a D3D object with the **IID** of **ID3DDestructionNotifier**.
+
+
+## -inheritance
+
+The <b xmlns:loc="http://microsoft.com/wdcml/l10n">ID3DDestructionNotifier</b> interface inherits from the <a href="https://docs.microsoft.com/windows/desktop/api/unknwn/nn-unknwn-iunknown">IUnknown</a> interface. <b>ID3DInclude</b> also has these types of members:
+<ul>
+<li><a href="https://docs.microsoft.com/">Methods</a></li>
+</ul>
+
+## -members
+
+The <b>ID3DDestructionNotifier</b> interface has these methods.
+<table class="members" id="memberListMethods">
+<tr>
+<th align="left" width="37%">Method</th>
+<th align="left" width="63%">Description</th>
+</tr>
+<tr data="declared;">
+<td align="left" width="37%">
+<a href="windows/desktop/api/d3dcommon/nf-d3dcommon-id3ddestructionnotifier-registerdestructioncallback">RegisterDestructionCallback</a>
+</td>
+<td align="left" width="63%">
+Registers a user-defined callback to be invoked when the current object is destroyed.
+
+</td>
+</tr>
+<tr data="declared;">
+<td align="left" width="37%">
+<a href="windows/desktop/api/d3dcommon/nf-d3dcommon-id3ddestructionnotifier-unregisterdestructioncallback">UnregisterDestructionCallback</a>
+</td>
+<td align="left" width="63%">
+Unregisters a user-defined callback that was previously registered.
+
+</td>
+</tr>
+</table> 
+
+
+## -remarks
+
+The <b>ID3DDestructionNotifier</b> can be used to track resources which are being unexpectedly released early, or providing a log of object disposal.
```

Comment:
Note that destruction can also be tracked by using `ID3D11DeviceChild::SetPrivateDataInterface` or `ID3D12Object::SetPrivateDataInterface`, and using the provided object's `Release()` implementation to track destruction. The key difference between that object and this is ordering of destruction for intertwined objects.

For example, a D3D11 view holds an internal reference on a D3D11 resource. Using the `SetPrivateDataInterface` approach, if the resource is only being kept alive by the view, then releasing the view will result in the resource being destroyed first (as the view's internal reference to the resource is released before destroying the private data container).

The destruction notifier is specifically designed to explicitly order destruction of objects so that immutable object relationships, such as a view -> resource, are still valid at the time the notifiers are invoked.